Add Support for IGMP snooping on-off in eos_l3ls_evpn

titom73 opened the pull request:
Add a key to support to enable/disable igmp-snopping per SVI in tenant
section.

If variable is unset, we apply default EOS config (enable). Then we follow user definition

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] New feature (non-breaking change which adds functionality)

## Related Issue(s)

Fixes #315

## Component(s) name

`eos_l3ls_evpn`

## Proposed changes
<!--- Describe your changes in detail -->

```yaml
l3leaf:
  node_groups:
    DC1_LEAF2:
      bgp_as: 65102
      igmp_snooping_enabled: false

[ ... output trunctaed ... ]

l2leaf:
  defaults:
    igmp_snooping_enabled: false

[ ... output trunctaed ... ]

tenants:
  Tenant_A:
    vrfs:
      TENANT_A_PROJECT01:
        svis:
          110:
            igmp_snooping_enabled: false
  Tenant_B:
    l2vlans:
      201:
        igmp_snooping_enabled: false
```

## How to test

In group_vars:

```yaml
l3leaf:
  defaults:
    # virtual router mac for VNIs assigned to Leaf switches
    # format: xx:xx:xx:xx:xx:xx
    virtual_router_mac_address: 00:1c:73:00:dc:01
    platform: vEOS-LAB
    bgp_as: 65100
    spines: [DC1-SPINE1, DC1-SPINE2]
    uplink_to_spine_interfaces: [Ethernet1, Ethernet2]
    mlag_interfaces: [Ethernet3, Ethernet4]
    spanning_tree_priority: 4096
    spanning_tree_mode: mstp
  node_groups:
    DC1_LEAF1:
      bgp_as: 65101
      filter:
        tenants: [ 'Tenant_A', 'Tenant_B' ]
        tags: [ 'POD01', 'DC1' ]
        # tags: [ border ]
      nodes:
        DC1-LEAF1A:
          id: 1
          mgmt_ip: 10.73.255.111/24
          spine_interfaces: [Ethernet1, Ethernet1]
        DC1-LEAF1B:
          id: 2
          mgmt_ip: 10.73.255.112/24
          spine_interfaces: [Ethernet2, Ethernet2]
    DC1_LEAF2:
      bgp_as: 65102
      igmp_snooping_enabled: false
      filter:
        tenants: [ 'Tenant_A', 'Tenant_B' ]
        tags: [ 'POD02', 'DC1' ]
        # tags: [ border ]
      nodes:
        DC1-LEAF2A:
          id: 3
          mgmt_ip: 10.73.255.113/24
          spine_interfaces: [Ethernet3, Ethernet3]
        DC1-LEAF2B:
          id: 4
          mgmt_ip: 10.73.255.114/24
          spine_interfaces: [Ethernet4, Ethernet4]

# DC1 Tenants Networks
# Documentation of Tenant specific information - Vlans/VRFs
tenants:
  # Tenant A Specific Information - VRFs / VLANs
  Tenant_A:
    mac_vrf_vni_base: 10000
    vrfs:
      TENANT_A_PROJECT01:
        vrf_vni: 11
        svis:
          110:
            name: 'PR01-DMZ'
            tags: [DC1]
            enabled: true
            ip_address_virtual: 10.1.10.254/24
            igmp_snooping: false
          111:
            name: 'PR01-TRUST'
            tags: [POD02]
            enabled: true
            ip_address_virtual: 10.1.11.254/24
  Tenant_B:
    mac_vrf_vni_base: 20000
    l2vlans:
      201:
        name: 'B-ELAN-201'
        tags: [DC1]
        igmp_snooping: false
```

Check output on EOS:

```
DC1-LEAF1A(config-if-Vl101)#show ip igmp snooping vlan 110
   Global IGMP Snooping configuration:
-------------------------------------------
IGMP snooping                  : Enabled
IGMPv2 immediate leave         : Enabled
Robustness variable            : 2
Report flooding                : Disabled

Vlan 110 :
----------
IGMP snooping                  : Disabled
IGMPv2 immediate leave         : Default
Multicast router learning mode : pim-dvmrp
IGMP max group limit           : No limit set
Recent attempt to exceed limit : No
Report flooding                : Default
IGMP snooping pruning active   : False
Flooding traffic to VLAN       : True
IGMP snooping proxy active     : False
```

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly
- [x] I have read the [**CONTRIBUTING**](https://www.avd.sh/docs/contributing/) document.
- [x] All new and existing tests passed ([`pre-commit`](https://www.avd.sh/docs/installation/development/#python-virtual-environment), `make linting` and `make sanity-lint`).

titom73 committed:
Add Support for IGMP snooping on-off in eos_l3ls_evpn #315

Add a key to support to enable/disable igmp-snopping per SVI in tenant
section.

```
tenants:
  Tenant_A:
    vrfs:
      TENANT_A_PROJECT01:
        svis:
          110:
            ip_igmp_snooping: false
```

If variable is unset, we apply default EOS config (enable). Then we follow user definition

ClausHolbechArista commented:
For the use case where we need this, we will need to disable igmp snooping globally. Could you add this as well, or should we submit a separate issue?

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/tenant_evpn_vxlan/leaf-tenant-igmp-snooping.j2:
```diff
@@ -0,0 +1,37 @@
+{# Leaf tenant IGMP snooping #}
```
IGMP snooping is a L2 functionality, so we should support this for both SVIs and L2 VLANs and implement on both L2LEAF and L3LEAF.

titom73 replied:
Activated with commit aae1988

titom73 committed:
Add igmp-snopping for l2vlans #315

titom73 committed:
Add IGMP-snooping support for L2LEAFs #315

titom73 committed:
Update documentation for igmp-snooping configuration

titom73 commented:
I will add a specific key to activate/deactivate globally. Just searching for the correct variable to use.

titom73 committed:
Add a default variable to deactivate igmp-snooping fabric wise #315

titom73 committed:
Implement support of global IGMP snooping enable/disable #315

Add a new key in data model to support option to activate or deactivate
igmp snooping at device level.

Data input:
-----------

```yaml
ip_igmp_snooping:
  globally_enabled: False
```

Output:
-------

```
!
no ip igmp snooping
```

titom73 committed:
Implement logic to disable igmp snooping at device level #315

Implement logic to enable/disable igmp-snooping per MLAG cluster.

```
l3leaf:
  defaults:
    # Activate or deactivate IGMP snooping for all l3leaf devices | Optional default is true
    igmp_snooping_enabled: < true | false >
  MLAG-NODE-GROUP:
    igmp_snooping_enabled: < true | false >
```

If set to true, user can still configure IGMP-snooping per VNI
If set to fals, it disable igmp-snooping globally regardless VNI
settings.

titom73 committed:
Update IGMP variables for VNI

titom73 commented:
Data model has new entries for IGMP in `eos_l3ls_evpn` and a new one under `ip_igmp_snooping` for `eos_cli_config_gen`

cc @carlbuchmann and @aphillipps for data-model review

titom73 committed:
Extend igmp-snooping trigger #315

Implement full logic for igmp-snooping trigger

global: false
per-vni trigger to activate

global: true
per-vni trigger to deactivate

titom73 committed:
Remove scenario igmp false and igmp for svi true #315

Since EOS does not support IGMP globally disabled and enabled for 1 SVI
this scenario has been removed from `eos_l3ls_evpn`

Configuration:
--------------

```
DC1-LEAF1A(config)#no ip igmp snooping
DC1-LEAF1A(config)#ip igmp snooping vlan 110
```

Output:
-------

```
DC1-LEAF1A#show ip igmp snooping
   Global IGMP Snooping configuration:
-------------------------------------------
IGMP snooping                  : Disabled
IGMPv2 immediate leave         : Enabled
Robustness variable            : 2
Report flooding                : Disabled

Vlan 110 :
----------
IGMP snooping                  : Disabled
IGMPv2 immediate leave         : Default

```

titom73 committed:
Fix incorrect field for igmp_snooping_enabled in doc

titom73 commented:
Ready for review and testing

carlbuchmann reviewed (approved):
tested with avd-validation lab - LGTM!
